SFH bug fix, rewrite, and basic tests

maxbriel opened the pull request:
## Rewrite of star formation histories
This is a rewrite of the star formation histories to different classes that allows for "easier" adding of new SFHs. To this end, I've created the following additional classes:
1. An abstract SFHBase class defining the required components for a SFH.
2. A MadauBase class, which use the Madau&Dickinson 2014 CSFRD formulation, and Madau&Fragos 2017 (mean) metallicity evolution over redshift.

Available SFHs (in brackets is what it inherits from)
1. MadauDickinson14 (MadauBase)
2. MadauFragos17 (MadauBase)
3. Neijssel19 (MadauBase; metallicity overwritten)
4. IllustrisTNG (SFHBase)

**New SFHs:**

5. Chruslinska21 (+ individual submodels) (SFHBase)
6. Fujimoto24  (MadauBase)
7. Zalava21 (MadauBase)

Calling a class with the redshift and metallicity bins will return the CSFRD per redshift bin per metallicity bin.
For an example, see `SFR_per_Z_at_z(z, met_bins, MODEL)`.

These models can be called with `get_SFH_model` using the MODEL parameter. Adding the model here allows it to be used in the cosmic_weights calculation.

## Bug fix

While working on this, I noticed 2 bugs in the fraction SFR per metallicity calculation.
1. The total fraction summed over metallicity bins didn't always sum to 1. _This is caused by the wrong boundary being used for the upper bin metallicity bin. To include the weight in the final bin, it requires integration from the lower edge of the upper bin to our max integration value. However, the higher edge of the upper bin was used. This removed the material that was originally in the upper bin. This gave incorrect value and could lead to a ~15% decrease in high metallicity star formation!!_
2. The normalisation was not adapted per redshift value. _This was caused by only using the normalisation of the first z value only. Since the metallicity distribution changes over redshift, we need to normalise using the upper bound of the metallicity distribution at each redshift value._

## Tests
I've added a few basic tests to check that:
- the fSFR always sums to 1
- if the correct models are loaded
- if  `std_log_metallicity_dist` is giving back the right values
- if the `__call__` method of the classes returns the right multiplied values.

These are not meant to be complete tests but should provide a basis to build on. I do not have the time available to focus on these though.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    def __init__(self, MODEL):
+        self.MODEL = MODEL
+        # Automatically attach all model parameters as attributes
+        for key, value in MODEL.items():
```
Should we check/exclude, that MODEL has no key called MODEL, which would overwrite the MODEL?

maxbriel replied:
I don't think it's necessary.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        self.MODEL = MODEL
+        # Automatically attach all model parameters as attributes
+        for key, value in MODEL.items():
+            setattr(self, key, value)
```
Should we make some asserts here, that later assumed attributes are defined, e.g. `self.sigma`?

maxbriel replied:
These are different per sub-class, so cannot be checked here.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        for key, value in MODEL.items():
+            setattr(self, key, value)
+
+    @abstractmethod
+    def CSFRD(self, z):
+        """Compute the cosmic star formation rate density."""
+        pass
+
+    @abstractmethod
+    def mean_metallicity(self, z):
+        """Return the mean metallicity at redshift z."""
+        pass
+        
+    @abstractmethod
+    def fSFR(self, z, metallicity_bins):
+        """Return the fractional SFR as a function of redshift and metallicity bins."""
+        pass
```
Additionally, we should define the parameters in the docstrings.

maxbriel replied:
I've added the parameters and information and decided to explicitly mention that they need to be implemented in the subclass.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        array
+            Star formation history per metallicity bin at the given redshift(s).
+        '''
+        return self.CSFRD(z)[:, np.newaxis] * self.fSFR(z, met_bins)
```
This will always cause and error in the base class. Because the called functions will return `None`.

maxbriel replied:
One should not use an abstract base class on it's own. That's why this is an abstractclass and other calsses inherit from it.
This can only fail if in a sub-class CSFRD and fSFR are defined in a wrong way, which is developer dependent.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        pass
+
+    @abstractmethod
+    def mean_metallicity(self, z):
```
Why is this function requested, while not used?

maxbriel replied:
This forces each sub-class to contain this method. Same with the CSFRD and fSFR. 

These 3 functions don't do anything, except for sub-classes to contain them.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        float or array
+            The mean metallicity at the given redshift.
+        '''
+        return 10 ** (0.153 - 0.074 * z ** 1.34) * Zsun
```
Is that formula valid for any updated Zsun or does it rely on the Asplund+09 value? In the second case, we should have the Asplund value hard coded instead of using the constant for the solar metallicity.

maxbriel replied:
This is what was originally implemented. I'm not certain on this validity. Feel free to verify.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        
+        
+        # Use the first value of mu for normalisation
+        norm = stats.norm.cdf(np.log10(self.Z_max), mu_array, sigma)
```
Should we better calculate `Z_max` here to be the maximum of `metallicity_bins`, otherwise the normalization may not work.

maxbriel replied:
This way the normalisation works by limiting the metallicity distribution to `Z_max` at all redshift. This is why `norm` is unique at each redshift bin.
So at each redshift, the metallicity distribution is limited to `Z_max` (which is user given). Since we divide each distribution range by `norm`, the sum over metallicity bins should add up to 1 at each redshift.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    def mean_metallicity(self, z):
+        return 0.035 * 10 ** (-0.23 * z)
+    
+    # overwrite std_log_metallicity_dist method of MadauBase
```
Looks like here is a function missing, which overwrites the `std_log_metallicity_dist`.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        mu = np.log(self.mean_metallicity(z)) - sigma**2 / 2.0
+        # renormalisation constant
+        norm = stats.norm.cdf(np.log(self.Z_max), mu, sigma)
+        fSFR = np.empty((len(z), len(metallicity_bins) - 1))
```
Do we need that? Won't the next line create an ndarray of the correct shape?

maxbriel replied:
I think that might be right, but it shouldn't really matter for performance/outcome.

mkruckow replied:
This was more for readability of the code. Keeping unnecessary stuff in there may cause other coders to conserve it forever.

maxbriel replied:
Sure! I ended up generalising the calculation for the distribution across the metallicity bins, which made it necessary to keep the initialisation, but I changed them to np.zeros().

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        return np.load(os.path.join(PATH_TO_POSYDON_DATA, "SFR/IllustrisTNG.npz"))
+    
+    def CSFRD(self, z):
+        SFR_interp = interp1d(self.redshifts, self.SFR_data)
```
Just to crosscheck: the SFR in the IllustrisTNG.npz is the star formation rate density and not just the star formation rate? Maybe we should call the `self.SFR_data` already such that it makes clear it is already a density.

maxbriel replied:
it is a density yes.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    def mean_metallicity(self, z):
+        out = np.zeros_like(self.redshifts)
+        for i in range(len(out)):
+            if np.sum(self.M[i, :]) == 0:
+                out[i] = 0
+            else:
+                out[i] = np.average(self.Z, weights=self.M[i, :])
```
Can't we simplify this, as `out[i]` is already 0.
Beside that, do we really want the mean metallicity to be 0 in cases, we can't normalize it or should we set it to `np.nan` instead?

maxbriel replied:
I've set it as `np.nan`. This is more for comparison between the different SFH metallicity evolutions.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    def fSFR(self, z, metallicity_bins):
+        # only use data within the metallicity bounds (no lower bound)
+        Z_max_mask = self.Z <= self.Z_max
+        redshift_indices = np.array([np.where(self.redshifts <= i)[0][0] for i in z])
```
Minor: I'd suggest to not use `i` as the loop variable, usually one assumes that a looping `i` is an integer. I'd suggest to name the loop variable `redshift`.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        # as used in the calculations - do not change
+        # This is the metallicity bin edges used in the Chruslinska+21 calculations
+        FOH_min, FOH_max = 5.3, 9.7
+        self.FOH_bins = np.linspace(FOH_min,FOH_max, 200)
```
```suggestion
        self.FOH_bins = np.linspace(FOH_min, FOH_max, 200)
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        # This is the metallicity bin edges used in the Chruslinska+21 calculations
+        FOH_min, FOH_max = 5.3, 9.7
+        self.FOH_bins = np.linspace(FOH_min,FOH_max, 200)
+        self.dFOH=self.FOH_bins[1]-self.FOH_bins[0]
```
```suggestion
        self.dFOH = self.FOH_bins[1] - self.FOH_bins[0]
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    def _FOH_to_Z(self, FOH):
+        # scalings from Chruslinksa+21
+        if self.Z_solar_scaling == 'Asplund09':
+            Zsun, FOHsun = [0.0134, 8.69]
```
```suggestion
            Zsun = 0.0134
            FOHsun = 8.69
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        if self.Z_solar_scaling == 'Asplund09':
+            Zsun, FOHsun = [0.0134, 8.69]
+        elif self.Z_solar_scaling == 'AndersGrevesse89':
+            Zsun,FOHsun = [0.017, 8.83]
```
```suggestion
            Zsun = 0.017
            FOHsun = 8.83
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        elif self.Z_solar_scaling == 'AndersGrevesse89':
+            Zsun,FOHsun = [0.017, 8.83]
+        elif self.Z_solar_scaling == 'GrevesseSauval98':
+            Zsun,FOHsun = [0.0201, 8.93]
```
```suggestion
            Zsun = 0.0201
            FOHsun = 8.93
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        else:
+            raise ValueError("Invalid Z_solar_scaling!")
+        logZ = np.log10(Zsun) + FOH - FOHsun
+        ZZ=10**logZ
+        return ZZ
```
```suggestion
        return 10**logZ
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        elif self.Z_solar_scaling == 'Villante14':
+            Zsun,FOHsun = [0.019, 8.85]
+        else:
+            raise ValueError("Invalid Z_solar_scaling!")
```
I'd suggest to add the allowed values in the error message.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        
+        Parameters
+        ----------
+        z : float or array-like
+            Cosmological redshift.
+        metallicity_bins : array
+            Metallicity bins edges in absolute metallicity.
+        
+        Returns
+        -------
+        array
+            Fraction of the SFR in the given metallicity bin at the given redshift.
+        '''
+        # only use data within the metallicity bounds (no lower bound)
+        Z_max_mask = self.Z <= self.Z_max
+        redshift_indices = np.array([np.where(self.redshifts <= i)[0][0] for i in z])
+        Z_dist = self.SFR_data[:, Z_max_mask][redshift_indices]
+        fSFR = np.zeros((len(z), len(metallicity_bins) - 1))
+        
+        for i in range(len(z)):
+            if Z_dist[i].sum() == 0.0:
+                continue
+            else:
+                # Add a final point to the CDF and metallicities to ensure normalisation to 1
+                Z_dist_cdf = np.cumsum(Z_dist[i]) / Z_dist[i].sum()
+                Z_dist_cdf = np.append(Z_dist_cdf, 1)
+                Z_x_values = np.append(np.log10(self.Z[Z_max_mask]), 0)
+                Z_dist_cdf_interp = interp1d(Z_x_values, Z_dist_cdf)
+
+                fSFR[i, :] = (Z_dist_cdf_interp(np.log10(metallicity_bins[1:])) -
+                              Z_dist_cdf_interp(np.log10(metallicity_bins[:-1])))
+
+                if not self.select_one_met:
+                    if len(metallicity_bins) == 2:
+                        fSFR[i, 0] = 1
+                    else:
+                        fSFR[i, 0] = Z_dist_cdf_interp(np.log10(metallicity_bins[1]))
+                        fSFR[i, -1] = 1 - Z_dist_cdf_interp(np.log10(metallicity_bins[-2]))
+                        
+        return fSFR
```
This looks the same as lines 303-330, hence should that function get defined that way on the parent class?

maxbriel replied:
It is the same, but I decided against it here, because it would have required and extra class for "simulations" and it might change again for another simulation. So I didn't think it's necessary to create an abstract class.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        self._data_folder = os.path.join(PATH_TO_POSYDON_DATA, "SFR/Chruslinska+21")
+        _, self.redshifts, delta_T = self._load_redshift_data(verbose)
+        M = self._load_raw_data()
+        self.SFR_data = np.array( [M[ii]/(1e6*delta_T[ii]) for ii in range(len(delta_T))])/self.dFOH
```
It feels a bit strange, that you normalize here with the bin width, but later you always use `self.SFR_data*self.dFOH`, thus, won't it be better to save only the product for this class?

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+            self.SFR_data = tmp_data['SFRD_min'].values
+        elif self.sub_model == 'max':
+            self.SFR_data = tmp_data['SFRD_max'].values
+        else:
```
Shell we have another sub_model which uses the mean of min and max?

maxbriel replied:
This is the data available from the Authors', they have an "obscured/non-obscured" amounts and that's what their min/max stand for

mkruckow replied:
Ah, OK, could you add this information in a comment in the code? That others will know, that `min` and `max` actually refer to obscured and non-obscured?

maxbriel replied:
will do!

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    
+    Returns
+    -------
+    SFHBase
```
Strictly speaking it returns a child of SFHBase class.

mkruckow commented on posydon/popsyn/synthetic_population.py:
```diff
@@ -2038,10 +2035,6 @@ def calculate_cosmic_weights(self, SFH_identifier, MODEL_in=None):
         else:
-            for key in MODEL_in:
-                if key not in DEFAULT_MODEL:
-                    raise ValueError(key + " is not a valid parameter name!")
```
Instead of removing I'd throw a Warning here, that keys get ignored/might be unsupported.

mkruckow commented on posydon/popsyn/synthetic_population.py:
```diff
@@ -2083,23 +2076,12 @@ def calculate_cosmic_weights(self, SFH_identifier, MODEL_in=None):
-        )
-
+        # Not for every event!        
+        SFR_per_Z_at_z_birth = SFR_per_Z_at_z(z_birth, met_edges, MODEL)
```
Shell we pass here `MODEL` or `rates.MODEL` as before? I guess, the aim of using `rates.MODEL` was to ensure those are the values written to file before.

maxbriel replied:
changed.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,218 @@
+        # The sum over metallicity bins (axis=1) should be approximately 1 for each redshift
+        np.testing.assert_allclose(np.sum(fSFR, axis=1), np.ones(len(z)), atol=1e-6)
+    
+    def dummy_get_illustrisTNG_data(self, verbose=False):
+        return {
+            "SFR": np.array([1.0, 1.0, 1.0]),
+            "redshifts": np.array([0.0, 1.0, 2.0]),
+            "mets": np.linspace(0.001, 0.03, 10),
+            "M": np.ones((3, 10)),
+        }
+    
```
I'd suggest to move this definition inside the function, as it should not be callable outside of `test_fSFR_sum_is_one_illustris`.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,218 @@
+        dummy = DummySFH(model)
+        z = np.array([0.5, 1.0, 2.0])
+        met_bins = np.linspace(0.001, model["Z_max"], 10)
+        expected = 2 * dummy.fSFR(z, met_bins)
```
Why do we have a factor of `2`? Line 32 divides by the sum of `axis=1`, when by `len(met_bins)-1`, which should be 10.

maxbriel replied:
Because the CSFRD is filled with 2's :)

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,218 @@
+        np.testing.assert_allclose(result, expected, atol=1e-6)
+
+
+class TestGetSFHModel:
```
Shouldn't that be tested first? In case this fails the previous Tests would have failed, may not executing this here.

maxbriel replied:
Order is technically not guaranteed.

sgossage commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    else:
+        raise ValueError("Invalid SFR!")
+
+def SFR_per_Z_at_z(z, met_bins, MODEL):
```
```suggestion
def SFR_per_met_at_z(z, met_bins, MODEL):
```

I like the conciseness of the name, but `per_met_at_z` might be better for clarity of meaning in this case. What do you think?

maxbriel replied:
changed.

maxbriel commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+    and fractional SFR based on the chosen Madau parameterisation.
+    The specific parameters for CSFRD must be provided by subclasses.
+    """
+
```
the init is inherited from the SFHBase

maxbriel commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,596 @@
+        """Return the fractional SFR as a function of redshift and metallicity bins."""
+        pass
+
+    def std_log_metallicity_dist(self):
```
added

maxbriel commented:
General comments response:
> Inconsistent usage of single vs. double quote in the code, please make it concise.
Done

> All new functions should have proper docstrings.
I've added many new ones now.

> We should separate code/unit and validation tests.
I believe the code here only performs some basic unit tests and no validation tests. We do need the latter. 

> All PRs including new tests should add them to the continuous integration.
> The tests should cover all cases, e.g. "select_one_met": True, very high or low Z_max,
> The tests should cover all the functions, e.g. _load_chruslinska_data
As mentioned, these tests are not meant to be complete. As such, I do not add them to the CI nor have I added full coverage/ all test cases. They should, however, cover issues I tried to solve. @elizabethteng these were the tests I mentioned during the meeting.

> Please call array always by their correct name np.array or ndarray.
This is unclear. I assume you refer to output array types in the docstring, which have been adapted.

 > Please check the code style, e.g. 80 characters limit.
I've checked it and allowed some leniency with the maximum number if it inhibited readability.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,891 @@
+            The calculated star formation rate at the given redshift(s) and
+            metallicity bins in Msun/yr.
+
+        Raises:
+            NotImplementedError: If the subclass does not implement this method.
```
Could you please add this to the other abstract methods as well?

maxbriel replied:
yes!

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,891 @@
+        ndarray
+            Fraction of the SFR in the given metallicity bin at the given redshift.
+        '''
+        fSFR = (np.array(cdf_func(metallicity_bins[1:]))
+                    - np.array(cdf_func(metallicity_bins[:-1])))
```
```suggestion
        fSFR = (np.array(cdf_func(metallicity_bins[1:]))
                - np.array(cdf_func(metallicity_bins[:-1])))
```

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,6 +36,891 @@
+                print("Warning: Z_min is larger than the lowest metallicity bin.")
+                fSFR[0] = 0.0
+            
+        if self.normalise:
+            fSFR /= np.sum(fSFR)
```
Out of curiosity: What is the future use case for this? Currently `self.normalise` is always `False`.

maxbriel replied:
It's not if it's given by the user in the MODEL. Then it's activated and normalised.

mkruckow replied:
Sure, my question was, which model does that?

maxbriel replied:
The latter is correct. The user can give parameters as part of the initialisation of the pre-defined SFH classes. As the code shows, MODEL is a required input from the SFHBase class. This is where the user can give `normalise` as an option.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,904 @@
+                If True, the SFR is calculated for a single metallicity bin.
+        """
+        
+        self.Z_max = None
+        self.Z_min = None
+        self.normalise = False
+        
```
```suggestion
```
This looks like a duplicate of calling the `__init__` of the parent class.

elizabethteng commented:
Sorry for the delay in review; I just finished reading through the changes. I can't evaluate it from a physics standpoint but the code generally looks good to me. I did have a bit of trouble understanding how the abstractmethods in star_formation_history.py work, could you explain a bit?

For the tests, I'm a little confused by how the tests for TestSFHBase are actually instantiating the SFHBase code. It looks like the classes you are constructing inherit from SFHBase, but is that the same as creating an instance?

I think you might be able to construct a fixture for ConcreteSFH so you don't have to duplicate the code multiple times under different functions.

maxbriel commented:
@elizabethteng

I found it hard to test an abstract class since it can't be used independently. I thought that checking if a barebones class inherits correctly would be useful as a test, because you cannot use the SFHBase directly without CSFRD, mean_metallicity and fSFR function definitions.

For the use case where it can be done, I moved the ConcreteSFH into a fixture, but there are a few different versions that I've kept separate.

elizabethteng commented:
that makes sense, thanks!

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+import numpy as np
```
Could you please add a module docstring and the author information like for other code files?

mkruckow commented:
Could you please check to capture/check the warnings in your test (DeprecationWarnings from other modules you can skip as they are not caused by us and only hint what we need to check when updating to newer versions of the external modules)? This is from the pytest output:

> posydon/unit_tests/popsyn/test_star_formation_history.py::TestMadauBase::test_fsfr
>   /home/runner/work/POSYDON/POSYDON/posydon/popsyn/star_formation_history.py:322: RuntimeWarning: divide by zero encountered in log10
>     cdf_func = lambda x: stats.norm.cdf(np.log10(x), mean, sigma)
> 
> posydon/unit_tests/popsyn/test_star_formation_history.py::TestMadauBase::test_fsfr
>   /home/runner/work/POSYDON/POSYDON/posydon/popsyn/star_formation_history.py:181: UserWarning: Z_min is larger than the lowest metallicity bin.
>     Pwarn('Z_min is larger than the lowest metallicity bin.')

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+                return z
+        
+        model_dict = {"Z_max": 0.03}
+        with pytest.raises(TypeError):
+            IncompleteSFH1(model_dict)
```
Do the TypeError tell the details, what function is missing? If so, shouldn't we check for that?

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        expected = np.array([0.01, 0.01, 0.98])
+        np.testing.assert_allclose(result, expected)
+        
+        # Test with normalization
+        sfh.normalise = True
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        np.testing.assert_allclose(np.sum(result), 1.0)
```
Does it really test the normalization? The unnormalized version would already sum up to 1=0.01+0.01+0.98. How about using a Z_max=0.5, which changes 0.98 to 0.48 in the unnormalized expectation?

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,899 @@
+            if self.Z_max >= metallicity_bins[-1]:
+                fSFR[-1] = cdf_func(self.Z_max) - cdf_func(metallicity_bins[-2])
+            else:
+                print(f"Z_max is smaller than the highest metallicity bin.")
```
```suggestion
```
It's a duplicate of the warning below.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        # Test model dict warning
+        model_dict = {"Z_max": 0.02, "Z_min": 0.0}
+        sfh = ConcreteSFH(model_dict)
+        with pytest.warns(UserWarning):
```
It would be good to check for the content of the warning, to ensure the correct warning it thrown.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        np.testing.assert_allclose(np.sum(result), 1.0)
+
+        # test no Z_min/Z_max set
+        model_dict = {}
+        sfh = ConcreteSFH(model_dict)
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        np.testing.assert_allclose(result, 0.01 * np.ones(3))
+        
+        # Test model dict warning
+        model_dict = {"Z_max": 0.02, "Z_min": 0.0}
+        sfh = ConcreteSFH(model_dict)
+        with pytest.warns(UserWarning):
+            result = sfh._distribute_cdf(cdf_func, met_edges)
+        
+        # Test with different model dicts
+        model_dict = {"Z_max": 1, "Z_min": 0.015}
+        sfh = ConcreteSFH(model_dict)
+        met_edges = np.array([0.3, 0.6, 0.9])
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        expected = np.array([0.585, 0.4])
+        np.testing.assert_allclose(result, expected)
+        
+        # Test with normalise
+        sfh.normalise = True
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        np.testing.assert_allclose(np.sum(result), 1.0)
+        
+        # restrict the upper bound
+        model_dict = {"Z_max": 0.95, "Z_min": 0.2}
+        sfh = ConcreteSFH(model_dict)
+        met_edges = np.array([0.3, 0.6, 0.9])
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        expected = np.array([0.4, 0.35])
+        np.testing.assert_allclose(result, expected)
+        
+        # Test with normalise
+        sfh.normalise = True
+        result = sfh._distribute_cdf(cdf_func, met_edges)
+        np.testing.assert_allclose(np.sum(result), 1.0)
```
Should we use loops for recurrent tests?

maxbriel replied:
I've implemented it as a parameterisation.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+
+    def test_call_method(self):
+        """Test the __call__ method."""
+        class ConcreteSFH(SFHBase):
```
May name it different than the other version giving a float. So, may call them ConcreteSFH_float, ConcreteSFH_ndarray?

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        model_dict = {"Z_max": 0.03}
+        sfh = ConcreteSFH(model_dict)
+        
+        z = np.array([0.5, 1.0])
```
The value of z doesn't matter here. Thus, this test may let people think that having the same shape is required and tested.
Is there a way to ensure that the input and output of our abstract methods are checked to have the same type and shape?

maxbriel replied:
Test adapted to use z and met_edges

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,899 @@
+        if "sigma" not in MODEL:
+            raise ValueError("sigma not given!")
+        super().__init__(MODEL)
+        self.CSFRD_params = None
```
Is there a reason to now allow the `CSFRD_params` to be set in the `MODEL`?

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+            madau.std_log_metallicity_dist()
+        assert "Unknown sigma choice!" in str(excinfo.value)
+        
+        # Test invalid type
+        model_dict = {"sigma": 1, "Z_max": 0.03}
+        madau = self.ConcreteMadau(model_dict)
+        with pytest.raises(ValueError) as excinfo:
+            madau.std_log_metallicity_dist()
+        assert "Invalid sigma value" in str(excinfo.value)
```
Shouldn't we test with a different type and allow integers?
`elif isinstance(sigma, float):` -> `elif isinstance(sigma, (float, int)):` if you like you can cast it to a float on the return statement.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        def __init__(self, MODEL):
+            super().__init__(MODEL)
+            self.CSFRD_params = {
+                "a": 0.01,
+                "b": 2.6, 
+                "c": 3.2,
+                "d": 6.2
```
Should we get a fixture with those values, so that we can reuse them instead of hard coding them?

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        # Test with array of values
+        z_array = np.array([0.0, 1.0, 2.0])
+        result = madau.CSFRD(z_array)
+        expected = np.array([
+            0.01 * 1**2.6 / (1 + (1/3.2)**6.2),  # z=0 
+            0.01 * 2**2.6 / (1 + (2/3.2)**6.2),  # z=1
+            0.01 * 3**2.6 / (1 + (3/3.2)**6.2)   # z=2
+        ])
```
May use the same style as in the following test(s).

maxbriel replied:
made a tmp function instead.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        
+        # THIS IS A VALIDATION TEST
+        from scipy.stats import norm
+        mean0, mean1 = (np.log10(madau.mean_metallicity(z)) 
+                  - model_dict['sigma']**2 * np.log(10) / 2)
+        print(mean0, mean1)
```
```suggestion
        mean0, mean1 = (np.log10(madau.mean_metallicity(z)) 
                        - model_dict['sigma']**2 * np.log(10) / 2)
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        
+        expected1 = np.array(
+            # integral from 0.001 to 0.01; Z_min = lowest bin edge
+            [norm.cdf(np.log10(0.01), mean0, model_dict['sigma'])
+             - norm.cdf(np.log10(0.001), mean0, model_dict['sigma']),
+            # integral from 0.01 to 0.02
+             (norm.cdf(np.log10(0.02), mean0, model_dict['sigma'])
+              - norm.cdf(np.log10(0.01), mean0, model_dict['sigma'])),
+            # integral from 0.02 to 1
+             (norm.cdf(np.log10(1), mean0, model_dict['sigma'])
+              - norm.cdf(np.log10(0.02), mean0, model_dict['sigma']))],)
```
```suggestion
            [(norm.cdf(np.log10(0.01), mean0, model_dict['sigma'])
              - norm.cdf(np.log10(0.001), mean0, model_dict['sigma'])),
            # integral from 0.01 to 0.02
             (norm.cdf(np.log10(0.02), mean0, model_dict['sigma'])
              - norm.cdf(np.log10(0.01), mean0, model_dict['sigma'])),
            # integral from 0.02 to 1
             (norm.cdf(np.log10(1), mean0, model_dict['sigma'])
              - norm.cdf(np.log10(0.02), mean0, model_dict['sigma']))],)
```
Just for consistency.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        
+        expected2 = np.array(
+            # integral from 0.001 to 0.01;Z_min = lowest bin edge
+            [norm.cdf(np.log10(0.01), mean1, model_dict['sigma'])
+             - norm.cdf(np.log10(0.001), mean1, model_dict['sigma']),
+            # integral from 0.01 to 0.02
+             (norm.cdf(np.log10(0.02), mean1, model_dict['sigma'])
+              - norm.cdf(np.log10(0.01), mean1, model_dict['sigma'])),
+            # integral from 0.02 to 1
+             (norm.cdf(np.log10(1), mean1, model_dict['sigma'])
+              - norm.cdf(np.log10(0.02), mean1, model_dict['sigma']))],
+                      )
```
```suggestion
            [(norm.cdf(np.log10(0.01), mean1, model_dict['sigma'])
              - norm.cdf(np.log10(0.001), mean1, model_dict['sigma'])),
            # integral from 0.01 to 0.02
             (norm.cdf(np.log10(0.02), mean1, model_dict['sigma'])
              - norm.cdf(np.log10(0.01), mean1, model_dict['sigma'])),
            # integral from 0.02 to 1
             (norm.cdf(np.log10(1), mean1, model_dict['sigma'])
              - norm.cdf(np.log10(0.02), mean1, model_dict['sigma']))],)
```
Just for consistency.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        model_dict = {"sigma": 0.5, "Z_max": 0.3, "Z_min": 0, "normalise": True}
+        madau = self.ConcreteMadau(model_dict)
+        result = madau.fSFR(z, met_bins)
+        print(result)
```
```suggestion
```
cleanup

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        print(result)
+        np.testing.assert_allclose(np.sum(result, axis=1), np.ones(2))
+        
+        # Test with Z_min > 0.03
```
Actually Z_min=0.02<0.03:
```suggestion
        # Test with Z_min > met_bins[1]
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        madau = self.ConcreteMadau(model_dict)
+        result = madau.fSFR(z, met_bins)
+        print(result)
+        np.testing.assert_allclose(np.sum(result, axis=1), np.ones(2))
```
For consistency.
```suggestion
        expected = np.ones(2)
        np.testing.assert_allclose(np.sum(result, axis=1), expected)
```

May consider to infer the `2` from the result's shape.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        np.testing.assert_allclose(result, Z_interp)
+        
+        # Test empty mass array
+        flipped_masses[0] = np.zeros_like(flipped_masses[0])
```
Do you mean to change the data in the `illustris_model` object? If so, it would be better to write it like this and not use a reference.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        assert isinstance(madau, MadauBase)
+    
+    def test_csfrd_calculation(self):
+        """Test that CSFRD calculations match expected values"""
```
What do we test here? It looks like we retest the function of the base class `MadauBase`. Hence should we skip/move this test? (the same holds for other `test_csfrd_calculation`)

maxbriel replied:
Yeah. This is more of a validation test to see if the specific CSFRD are giving the right values, but I guess with `test_init_parameters`, they're not really necessary.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        model_dict = {"sigma": 0.5, "Z_max": 0.3, "normalise": True}
+        neijssel = Neijssel19(model_dict)
+        result = neijssel.fSFR(z, met_bins)
+        np.testing.assert_allclose(np.sum(result, axis=1), np.ones(2))
```
For consistency.
```suggestion
        expected = np.ones(2)
        np.testing.assert_allclose(np.sum(result, axis=1), expected)
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        }
+    
+    @pytest.fixture
+    def chruslinska_model(self, mock_chruslinska_data, monkeypatch):
```
```suggestion
    def chruslinska_model(self, mock_chruslinska_data):
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        result = chruslinska_model._FOH_to_Z(FOH_test)
+        
+        # Expected: 10^(log10(0.0134) + FOH - 8.69)
+        Zsun = 0.0134
+        FOHsun = 8.69
+        expected = 10**(np.log10(Zsun) + FOH_test - FOHsun)
```
For consistency.
```suggestion
        expected = 10**(np.log10(0.0134) + FOH_test - 8.69)
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        # Test at specific redshifts
+        z_values = np.array([0.0, 2.0, 4.0])
+        result = chruslinska_model.mean_metallicity(z_values)
+        print(result)
```
cleanup
```suggestion
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        # Should be an array of the same length as z_values
+        assert len(result) == len(z_values)
+        
+        # Should decrease with increasing redshift
```
That's only true for the mock data. Maybe we should state that in the comment. I just looked in the paper and didn't see a monotonic decrease and got confused.

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,899 @@
+        tmp_data = pd.read_csv(data_file,
+                               names=["redshift", "SFRD_min", "SFRD_max"],
+                               skiprows=1,
+                               sep="\s+")
```
I don't know what "\s+" is for. Python looks to not know it either: `invalid escape sequence '\s'`. I couldn't find the file `SFR/Zavala+21.txt`, hence I couldn't check what character is used.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+        zavala_max = Zavala21(model_dict)
+        assert zavala_max.sub_model == "max"
+        
+        # Test invalid sub_model
+        model_dict = {"sub_model": "invalid", "Z_max": 0.03, "sigma": 0.5}
+        with pytest.raises(ValueError) as excinfo:
+            zavala_invalid = Zavala21(model_dict)
+        assert "Invalid sub-model!" in str(excinfo.value)
```
This doesn't test the code, it only tests the mocked function. It would be better to not mock the `_load_zavala_data` but the `pd.read_csv` in there.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+    def test_illustris_tng_model(self, monkeypatch):
+        """Test that get_SFH_model returns IllustrisTNG instance."""
+        # Mock the data loading method
+        def mock_get_data(self, verbose=False):
+            # Return minimal mock data structure
+            return {
+                "SFR": np.array([0.1, 0.2, 0.3]),
+                "redshifts": np.array([0.0, 1.0, 2.0]),
+                "mets": np.array([0.001, 0.01, 0.02]),
+                "M": np.ones((3, 3))
+            }
+        
+        # Patch the data loading method
+        monkeypatch.setattr(IllustrisTNG, "_get_illustrisTNG_data", mock_get_data)
```
As you only want to check the object type, wouldn't it be simpler to mock the `__init__` of `IllustrisTNG` (and the others below)? If you like to replace only functions in the `__init__` we should add more checks that all the class needs to do in the initialization is done.

maxbriel replied:
That's why each model has an `__init__` test for its class.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+    def test_invalid_model(self):
+        """Test that get_SFH_model raises an error for an invalid model."""
+        
+        model_dict = {"SFR": "InvalidModel", "sigma": 0.5, "Z_max": 0.03}
```
Other parameters don't matter for a wrong SFR.
```suggestion
        model_dict = {"SFR": "InvalidModel"}
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1059 @@
+            model = get_SFH_model(model_dict)
+        assert "Invalid SFR!" in str(excinfo.value)
+
+class TestSFHUtilityFunctions:
+    """Tests for utility functions in the star_formation_history module."""
```
For consistency (within this test file): Should we move the tests for `get_SFH_model` in here as well? Or the other way around, making it a class for `SFR_per_met_at_z` only.

maxbriel replied:
Renamed instead.

mkruckow commented:
It looked to me that the old parameter `select_one_met` is still present in the docstrings, while I haven't seen any use of it anymore. Hence, should it get removed or used?

ZepeiX commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,899 @@
+            else:
+                print(f"Z_max is smaller than the highest metallicity bin.")
+                Pwarn('Z_max is smaller than the highest metallicity bin.')
+                fSFR[-1] = 0.0
```
allow the Z_max to be within the bin

maxbriel replied:
implemented.

maxbriel commented on posydon/popsyn/synthetic_population.py:
```diff
@@ -2038,11 +2035,7 @@ def calculate_cosmic_weights(self, SFH_identifier, MODEL_in=None):
-
-            # write the DEFAULT_MODEL with updates parameters to self.MODEL.
+            # write the DEFAULT_MODEL with updates parameters to MODEL.
             MODEL = DEFAULT_MODEL
```
This causes issues with  `Z_max` since it's set to 1.0 in the DEFAULT_MODEL. Adapt

maxbriel replied:
`Z_max` set to `None` in `DEFAULT_SFH_MODEL`.

sgossage commented on posydon/popsyn/synthetic_population.py:
```diff
@@ -2038,11 +2035,7 @@ def calculate_cosmic_weights(self, SFH_identifier, MODEL_in=None):
-                    raise ValueError(key + " is not a valid parameter name!")
-
-            # write the DEFAULT_MODEL with updates parameters to self.MODEL.
+            # write the DEFAULT_MODEL with updates parameters to MODEL.
```
```suggestion
            # write the DEFAULT_MODEL with updated parameters to MODEL.
```

sgossage commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,915 @@
+        
+        Adds the model parameters as attributes.
+        
+        Parameters
```
We should add descriptions for the other parameters `Z_max`, `Z_min`, `normalise`.

maxbriel replied:
I've also added them to the other SFH classes

maxbriel commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +38,899 @@
+    SFH = get_SFH_model(MODEL)
+    return SFH(z, met_bins)
+
+# TODO: No testing coverage for the following function, but should be added
```
Not part of this PR. Mentioned in PR #499

maxbriel replied:
I've mentioned on PR #499 that this test needs to be implemented.

maxbriel commented:
I've gone through all the comments and implemented most of them.
@mkruckow @sgossage Are the changes satisfactory?

mkruckow commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +37,974 @@
+            if self.Z_max >= metallicity_bins[-1]:
+                fSFR[-1] = cdf_func(self.Z_max) - cdf_func(metallicity_bins[-2])
+            else:
+                Pwarn('Z_max is smaller than the highest metallicity bin.')
```
Currently, this creates a UserWarning, should we change it into a POSYDONWarning or a subclass? (The same applies to other warnings in this file.)

maxbriel replied:
I've added a `SFHModelWarning` subclass and will create that.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1049 @@
+        
+        # THIS IS A VALIDATION TEST
+        from scipy.stats import norm
+        mean0, mean1 = (np.log10(madau.mean_metallicity(z)) 
+                  - model_dict['sigma']**2 * np.log(10) / 2)
```
Missed to change?
```suggestion
        mean0, mean1 = (np.log10(madau.mean_metallicity(z)) 
                        - model_dict['sigma']**2 * np.log(10) / 2)
```

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1049 @@
+        # Should be an array of the same length as z_values
+        assert len(result) == len(z_values)
+        
+        # Should decrease with increasing redshift in the 
```
```suggestion
        # Should decrease with increasing redshift in the test case
```

astroJeff commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +37,974 @@
+                raise ValueError("Z_min must be in absolute units! "
+                                 "It cannot be larger than 1!")
+        if self.Z_min is not None and self.Z_max is not None:
+            if self.Z_min >= self.Z_max:
```
We need to make sure there is a way to run a SFH with only one metallicity. An obvious way to do that is to set `Zmin` and `Zmax` to be equal. However, the `>=` comparison here will make that impossible.

maxbriel commented on posydon/popsyn/star_formation_history.py:
```diff
@@ -35,7 +37,974 @@
+        
+        # include material outside the metallicity bounds if requested
+        if self.Z_max is not None:
+            if self.Z_max >= metallicity_bins[-1]:
```
That is a good one. In principle, this takes the sorted array from `Rates().edges_metallciity_bins` which is sorted. But it does not verify or check here.

I will add a check to make sure it is sorted.

maxbriel commented:
@astroJeff @mkruckow  I've implemented your recent comments, except for checking the exact warning message. I just check if there's output.

mkruckow commented on posydon/unit_tests/popsyn/test_star_formation_history.py:
```diff
@@ -0,0 +1,1065 @@
+        ({"Z_max": 1, "Z_min": 0.015}, True, np.array([0.3, 0.6, 0.9]), 
+         None, None),
+        # Restrict upper bound
+        ({"Z_max": 0.95, "Z_min": 0.2}, False, np.array([0.3, 0.6, 0.9]), 
+         np.array([0.4, 0.35]), None),
+        # With normalization
+        ({"Z_max": 0.95, "Z_min": 0.2}, True, np.array([0.3, 0.6, 0.9]), 
+         None, None),
+        # Minimum in lowest bin
+        ({"Z_min": 0.25}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
+         np.array([0.05, 0.3, 0.3]), SFHModelWarning),
+        # Minimum higher than minimum bin
+        ({"Z_min": 0.35}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
+         np.array([0.0, 0.25, 0.3]), SFHModelWarning),
+        # Minimum in lowest bin and maximum
+        ({"Z_min": 0.25, "Z_max": 0.8}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
+         np.array([0.05, 0.3, 0.2]), SFHModelWarning),
+        # Minimum higher than minimum bin, narrow range
+        ({"Z_min": 0.35, "Z_max": 0.4}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
+         np.array([0.0, 0.05, 0.0]), SFHModelWarning),
+        # Minimum higher than minimum bin, medium range
+        ({"Z_min": 0.35, "Z_max": 0.65}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
+         np.array([0.0, 0.25, 0.05]), SFHModelWarning),
+    ])
+    def test_distribute_cdf(self, ConcreteSFH, model_dict, normalise, met_edges, expected, warning):
+        """Test the _distribute_cdf method with various scenarios."""
+        # Create a simple CDF function
+        cdf_func = lambda x: x
+        
+        # Create the SFH instance
+        sfh = ConcreteSFH(model_dict)
+        sfh.normalise = normalise
+        
+        # Test execution with or without warning check
+        if warning:
+            with pytest.warns(warning) as excinfo:
+                result = sfh._distribute_cdf(cdf_func, met_edges)
+            assert excinfo[0].message.args[0] is not None
+        else:
+            result = sfh._distribute_cdf(cdf_func, met_edges)
```
Sorry, I may confused you with putting two request in one comment, what I meant was:
```suggestion
    @pytest.mark.parametrize(
        "model_dict, normalise, met_edges, expected, warning", 
        [
        # Simple CDF with Z_max=1, Z_min=0.0
        ({"Z_max": 1, "Z_min": 0.0}, False, np.array([0.0, 0.01, 0.02, 0.03]), 
         np.array([0.01, 0.01, 0.98]), None),
        # No Z_min/Z_max set
        ({}, False, np.array([0.0, 0.01, 0.02, 0.03]), 
        0.01 * np.ones(3), None),
        # Model dict warning
        ({"Z_max": 0.02, "Z_min": 0.0}, False, np.array([0.0, 0.01, 0.02, 0.03]), 
         None, 'Z_max is smaller than the highest metallicity bin.'),
        # Different model dicts
        ({"Z_max": 1, "Z_min": 0.015}, False, np.array([0.3, 0.6, 0.9]), 
         np.array([0.585, 0.4]), None),
        # With normalization
        ({"Z_max": 1, "Z_min": 0.015}, True, np.array([0.3, 0.6, 0.9]), 
         None, None),
        # Restrict upper bound
        ({"Z_max": 0.95, "Z_min": 0.2}, False, np.array([0.3, 0.6, 0.9]), 
         np.array([0.4, 0.35]), None),
        # With normalization
        ({"Z_max": 0.95, "Z_min": 0.2}, True, np.array([0.3, 0.6, 0.9]), 
         None, None),
        # Minimum in lowest bin
        ({"Z_min": 0.25}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
         np.array([0.05, 0.3, 0.3]),
         'Z_min is larger than the lowest metallicity bin.'),
        # Minimum higher than minimum bin
        ({"Z_min": 0.35}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
         np.array([0.0, 0.25, 0.3]),
         'Z_min is larger than the lowest metallicity bin.'),
        # Minimum in lowest bin and maximum
        ({"Z_min": 0.25, "Z_max": 0.8}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
         np.array([0.05, 0.3, 0.2]),
         'Z_max is smaller than the highest metallicity bin.'),
        # Minimum higher than minimum bin, narrow range
        ({"Z_min": 0.35, "Z_max": 0.4}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
         np.array([0.0, 0.05, 0.0]),
         'Z_max is smaller than the highest metallicity bin.'),
        # Minimum higher than minimum bin, medium range
        ({"Z_min": 0.35, "Z_max": 0.65}, False, np.array([0.2, 0.3, 0.6, 0.9]), 
         np.array([0.0, 0.25, 0.05]),
         'Z_max is smaller than the highest metallicity bin.'),
    ])
    def test_distribute_cdf(self, ConcreteSFH, model_dict, normalise, met_edges, expected, warning):
        """Test the _distribute_cdf method with various scenarios."""
        # Create a simple CDF function
        cdf_func = lambda x: x
        
        # Create the SFH instance
        sfh = ConcreteSFH(model_dict)
        sfh.normalise = normalise
        
        # Test execution with or without warning check
        if warning is not None:
            with pytest.warns(SFHModelWarning, match=warning):
                result = sfh._distribute_cdf(cdf_func, met_edges)
        else:
            result = sfh._distribute_cdf(cdf_func, met_edges)
```

mkruckow reviewed (approved):
Looks ready to go for me.
